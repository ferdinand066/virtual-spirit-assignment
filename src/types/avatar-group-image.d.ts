export type AvatarGroupImageModel = {
  id: string;
  placeholder: string;
  imageUrl?: string | null;
};

export type AvatarSize = 'xs' | 'sm' | 'md' | 'lg';