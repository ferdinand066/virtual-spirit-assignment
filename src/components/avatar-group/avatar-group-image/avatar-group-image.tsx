import { useState } from "react";
import { AvatarGroupImageModel, AvatarSize } from "../../../types/avatar-group-image";
import styles from "../avatar-group.module.css";
import classNames from "classnames";

type AvatarGroupImageProps = {
  image: AvatarGroupImageModel;
  size: AvatarSize;
};

export default function AvatarGroupImage({
  image,
  size,
}: AvatarGroupImageProps) {
  const [isImageLoaded, setIsImageLoaded] = useState(true);

  const imageClasses = classNames(
    styles["avatar-group-image"],
    styles["avatar-group-container-child"],
    styles[`image-${size}`]
  );

  const placeholderClasses = classNames(
    styles["avatar-group-text"],
    styles["avatar-group-container-child"], 
    styles[`image-${size}`],
  );

  return (
    <>
      {isImageLoaded ? (
        <img
          role="avatar-image"
          key={image.id}
          src={image.imageUrl ?? ""}
          alt={image.placeholder}
          onLoad={() => setIsImageLoaded(true)}
          onError={() => setIsImageLoaded(false)}
          className={imageClasses}
        />
      ) : (
        <div
          role="avatar-image"
          className={placeholderClasses}
        >
          {image.placeholder}
        </div>
      )}
    </>
  );
}
