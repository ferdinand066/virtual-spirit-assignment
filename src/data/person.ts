export const persons = [
  {
    id: '1',
    name: 'Joseph Gonzalez',
    imageUrl: 'https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2574&q=80'
  }, 
  {
    id: '2',
    name: 'Calvin Luis',
    imageUrl: 'invalid-url'
  },
  {
    id: '3',
    name: 'Jonathan de Tristan',
    imageUrl: 'https://images.unsplash.com/photo-1500648767791-00dcc994a43e?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80'
  },
  {
    id: '4',
    name: 'Brendan Abdul',
    imageUrl: 'https://images.unsplash.com/photo-1591084728795-1149f32d9866?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2564&q=80'
  },
  {
    id: '5',
    name: 'Aadam Jake Valentino',
    imageUrl: 'https://images.unsplash.com/photo-1599834562135-b6fc90e642ca?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80'
  },
  {
    id: '6',
    name: 'Elizabeth Joel',
    imageUrl: 'https://images.unsplash.com/photo-1553514029-1318c9127859?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=928&q=80'
  },
  {
    id: '7',
    name: 'Michaela Rosemary',
    imageUrl: 'https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80'
  },
  {
    id: '8',
    name: 'Jane Josie Josephine',
    imageUrl: 'https://images.unsplash.com/photo-1534528741775-53994a69daeb?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=928&q=80'
  },
  {
    id: '9',
    name: 'Katy Keira',
    imageUrl: 'https://images.unsplash.com/photo-1438761681033-6461ffad8d80?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1740&q=80'
  },
  {
    id: '10',
    name: 'Hana Chelsea',
    imageUrl: 'https://images.unsplash.com/photo-1619418170841-afea868fdc58?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1740&q=80'
  }
]