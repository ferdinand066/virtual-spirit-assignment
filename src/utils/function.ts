const getNameInitial = (name: string) => {
  const words = name.split(" ");

  if (words.length === 1) return words[0].charAt(0);
  return words[0].charAt(0) + words[words.length - 1].charAt(0);
}

export { getNameInitial }