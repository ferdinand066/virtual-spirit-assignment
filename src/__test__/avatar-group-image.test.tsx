import AvatarGroupImage from "../components/avatar-group/avatar-group-image/avatar-group-image";
import { fireEvent, render, waitFor, screen, cleanup } from "@testing-library/react";
import { getNameInitial } from "../utils/function";
import { persons } from "../data/person";
import AvatarGroup from "../components/avatar-group/avatar-group";
import { AvatarSize } from "../types/avatar-group-image";


describe("Test Avatar Group Image Component", () => {
  test("should render an image if it loads successfully", () => {
    const image = {
      id: "1",
      imageUrl: 'https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2574&q=80',
      placeholder: getNameInitial('Joseph Gonzalez'),
    };

    const { getByRole } = render(
      <AvatarGroupImage image={image} size="xs" />
    );

    const component = getByRole("avatar-image");

    expect(component).toBeInTheDocument();

    expect(component.className.includes('avatar-group-image')).toBe(true);
    expect(component.className.includes('avatar-group-text')).toBe(false);

    cleanup();
  });

  test("should render a placeholder div if it fails to load the image", async () => {
    const image = {
      id: "1",
      imageUrl: "invalid-url",
      placeholder: getNameInitial('Joseph Gonzalez'),
    };

    render(
      <AvatarGroupImage image={image} size="xs" />
    );

    await waitFor(() => {
      const img = screen.getByRole('avatar-image');
      fireEvent.error(img);
    });

    const component = screen.getByRole("avatar-image");

    expect(component.className.includes('avatar-group-image')).toBe(false);
    expect(component.className.includes('avatar-group-text')).toBe(true);

    cleanup();
  });

  test("should render images with the same number according to the parameters given",  () => {
    const images = persons.map((person) => ({
      ...person,
      placeholder: getNameInitial(person.name)
    }));

    const maxLength = 2;

    render(
      <AvatarGroup images={images} maxLength={maxLength} size="xs" />
    );

    const avatarImageComponents = screen.getAllByRole("avatar-image");
    const totalRemainingPeopleComponent = screen.getByRole("total-remaining-people");

    expect(avatarImageComponents.length === maxLength).toBe(true);
    expect(totalRemainingPeopleComponent).toHaveTextContent(`+${images.length - maxLength}`);
    cleanup();

  });

  test("should render images with the same size according to the size parameter given", () => {
    const images = persons.map((person) => ({
      ...person,
      placeholder: getNameInitial(person.name)
    }));

    const sizes: AvatarSize[] = ['xs', 'sm', 'md', 'lg'];

    sizes.forEach(size => {
      render(
        <AvatarGroup images={images} maxLength={2} size={size} />
      );

      const avatarImageComponents = screen.getAllByRole("avatar-image");
      const filteredAvatarImageComponents = avatarImageComponents.filter((component) => component.className.includes(`image-${size}`));

      expect(filteredAvatarImageComponents.length).toBe(avatarImageComponents.length);
      cleanup();
    });

    cleanup();
  });

  test('should component match snapshots', () => {
    const images = persons.map((person) => ({
      ...person,
      placeholder: getNameInitial(person.name)
    }));

    const { container } = render(
      <AvatarGroup images={images} maxLength={2} size="xs" />
    );

    expect(container).toMatchSnapshot();
  });
});
