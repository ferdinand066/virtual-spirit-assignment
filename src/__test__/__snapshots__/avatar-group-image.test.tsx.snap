// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Avatar Group Image Component > should component match snapshots 1`] = `
<div>
  <div
    class="_avatar-group-container_cdb978"
  >
    <img
      alt="JG"
      class="_avatar-group-image_cdb978 _avatar-group-container-child_cdb978 _image-xs_cdb978"
      role="avatar-image"
      src="https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2574&q=80"
    />
    <img
      alt="CL"
      class="_avatar-group-image_cdb978 _avatar-group-container-child_cdb978 _image-xs_cdb978"
      role="avatar-image"
      src="invalid-url"
    />
    <div
      class="_avatar-group-text_cdb978 _image-xs_cdb978"
      role="total-remaining-people"
    >
      +8
    </div>
  </div>
</div>
`;
